resources/nersc_supervisor.sh

```diff
@@ -0,0 +1,248 @@
+      return 1
+    fi
+  logit 2 "launch $1"
+  if ! ${SBATCH}  -o "${NS_SLURM_LOG}/slurm-$1-%j.out" -q xfer -M escori -t 12:00:00 "${NS_SCRIPT_PATH}/$1.sh"
+    then
+      logit 0 "Submit of $1 failed"
+      return 1
+    fi
+  return 0
+}
+
+###
+# Check the slurm job activity information, count the incomplete entries
+
+function countIncomplete {
+  # No arguments
+  #
+  logit 2 'Entering countIncomplete'
+  twoweeksbefore=$(/usr/bin/date +%s | /usr/bin/awk '{printf("%s\n",strftime("%Y-%m-%d",$1));}')
+  # NOTA BENE:  The same job number can appear several times in the output.
+  # Therefore I have to sort -u the result.
+  # bash will treat the contents of "listing" as a single line, of course.
+  listing=$(${SACCT} -n -b -S "${twoweeksbefore}" -M escori -p | /usr/bin/awk '{split($0,b,"|");split(b[1],a,".");if(b[2]!="COMPLETED")print a[1],b[2],b[3],"|";}' |/usr/bin/sort -u)
+  if [[ "${listing}" == "" ]]
+    then
+        echo 0
+        return 0
+    fi
+  # Steve doesn't think I need to log the incomplete info
+  #logit 2 "${listing}"
+  # Count the number of not-completed, skipping the terminal "|" if any
+  echo "${listing}" | /usr/bin/awk '{n=split($0,a,"|");m=n-1;if(m<0)m=0;print m;}'
+  return 0
+}
+
+###################################################################
+# This is where the main work really begins.
+#############################################
+#
+
+# First find and set the logging level
+
+if [[ "$1" == "" ]]
+  then
+    NS_LOG_DETAIL=0
+  else
+    NS_LOG_DETAIL=$1
+  fi
+# Log (or not, depending on the level)
+logit 1 "Invocation"
+
+# Not used yet
+
+
+
+
+##################
+# Start the work #
+##################
+
+#if ! incomplete=$(countIncomplete)
+#  then
+#    exit 1
+#  fi
+
+###
+# How many slurm jobs are running right now?
+###
+if ! whatWeHave=$(getrunning)
+  then
+    exit 2
+  fi
+
+###
+# Loop over the list of expected job types
+# At the moment this info is hard-wired in NS_DESIRED, but
+# we should load from a json initialization file instead
+# Keeping track of activeJobs is future-proofing the system
+# I should count them all and compare with maxjobs before
+# submitting anything.
+###
+activeJobs=0
+for desired in ${NS_DESIRED}
+  do
+    class=$(echo "${desired}" | /usr/bin/awk '{split($1,a,":");print a[1];}')
+    expectedcount=$(echo "${desired}" | /usr/bin/awk '{split($1,a,":");print a[2];}')
+    count=0
+    for chunks in ${whatWeHave}
+      do
+         fc=$(echo "${chunks}" | awk -v var="${class}.sh" '{n=split($0,a,var);print n-1}')
+         count=$(( count + fc ))
+      done
+      activeJobs=$(( activeJobs + count ))
+    ###
+    # Do we need to do anything?
+    ###
+    if [[ ${count} -lt ${expectedcount} ]]
+      then
+        # only launch 1 at a time
+        if ! launch "${class}"
+          then exit 3; fi
+        logit 2 "Launching ${class} ${count} ${expectedcount}"
+      fi
+  done
+
+########
+# Done #
+########
+exit 0
```

Comment:
On line 189 you've got a logit for invocation.
Do you also want one here at the end for normal termination?

Reply:
Good point